api/envoy/extensions/filters/http/jwt_authn/v3/config.proto

```diff
@@ -452,6 +453,13 @@ message RemoteJwks {
   //
   //
   config.core.v3.RetryPolicy retry_policy = 4;
+
+  // Refetch JWKS if extracted JWT has no KID or a KID that does not match cached JWKS's KID.
+  // Async JWKS fetching is independent to this refetching process. New JWKS is fetched on
+  // demand and the cache is isolated to the fetched worker thread.
+  //
+  // If all cached JWKS have no KID, each request will trigger a JWKS refetch.
```

Comment:
This seems like it could wind up DoSing the HTTP server with repeated requests.  How about adding some exponential backoff here?

Also, what happens when there are multiple requests in parallel?  For example, consider the case where a request triggers a new fetch and then a second request starts while the fetch is still in flight.  Does that trigger another fetch, or is it smart enough to know that there should be only one fetch in flight at a time?  And if the latter, what is the behavior of the second request -- does it just get delayed waiting for the fetch to complete?

Reply:
> This seems like it could wind up DoSing the HTTP server with repeated requests. How about adding some exponential backoff here? 

That is the plan and I've added a TODO in the code to follow up with that. There are some optimizations from https://github.com/envoyproxy/envoy/issues/14557 that would be good to add:

- Update JWKS across all threads once a new JWKS is refetched. 
- Rate limiting these fetches.

This does require a reasonable amount of code addition, which is the reason for the minimal config driven change that would help with certain usecases linked to the above issue; Most IdPs add KIDs and this should work. This config/flow in the current state would be problematic in an environment where JWKS don't have KIDs as you mentioned, which is why I highlighted that issue in the docs. Anyways, if you think it's necessary to address the rate limiting as part of this PR, I can add it. 

> Also, what happens when there are multiple requests in parallel? For example, consider the case where a request triggers a new fetch and then a second request starts while the fetch is still in flight. Does that trigger another fetch, or is it smart enough to know that there should be only one fetch in flight at a time? And if the latter, what is the behavior of the second request -- does it just get delayed waiting for the fetch to complete?

It is the same behavior as how the filter would work as described under "async fetch" - "If this feature is not enabled:" [here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto.html#extensions-filters-http-jwt-authn-v3-remotejwks). In a worker thread, all requests including the initial one are paused until the JWKS is retrieved.

I've addressed the other comments above.

Reply:
I'd welcome other input here, but it does seem dangerous to me to trigger this kind of refresh based on data plane traffic without adding appropriate backoff, since it provides a vector for significant amplification of a DoS attack.  It seems like it would be safer to implement backoff first.

Also, when you say that this re-fetch is independent of the async refresh, I assume that means that if a re-fetch is triggered due to KID mismatch, that that refresh would be local to the thread, even though with async refreshing the JWKS is normally cached in the main thread?  That seems fairly counter-intuitive, since it effectively disables async refreshing.  Does it do that temporarily or permanently?  If the former, what criteria are used to switch back to the main thread copy?

Would it make sense instead to say that if async fetch is enabled, then a worker thread can trigger a reload on the main thread instead of doing its own per-thread fetch?  That would have the nice property that enabling async refresh always means that the fetches are done on the main thread and disabling it always means that they're done on the worker threads, regardless of the value of this new field.

Reply:
>  It seems like it would be safer to implement backoff first.

That's fair. I'll add the backoff. 
 
As for the async fetching, AFAIK the main thread retrieves the new JWKS and pushes it to all worker threads to be cached. I can modify the logic to accomplish something similar based on your suggestion. If async fetch is enabled, all worker threads should be updated with new JWKS. If not, just the local thread should be updated.

Reply:
Just to be clear, the reason for my suggestion wasn't just that all worker threads get updated -- that is certainly important, but I don't think it's the only concern here.  I think we also want (a) not to have multiple fetches happening in parallel in multiple threads and (b) the fetch we trigger here should reset the timer for the next async re-fetch in the main thread.